scripts/parse_tools/xml_tools.py

```diff
@@ -242,101 +243,355 @@ def read_xml_file(filename, logger=None):
+        ...       <scheme>pbl_scheme</scheme>
+        ...     </group>
+        ...   </suite>
+        ... </suites>
+        ... '''
+        >>> root = ET.fromstring(xml)
+        >>> expand_nested_suites(root, '/no/valid/path', logger)
+        Expanded nested suite 'microphysics_suite', group 'micro'
+        Expanded nested suite 'pbl_suite'
+        Removed nested suite 'microphysics_suite' from root element
+        Removed nested suite 'pbl_suite' from root element
+        >>> len(root.findall("suite"))  # Only one suite left
+        1
+        >>> suite = root.find("suite")
+        >>> suite.attrib.get("name")
+        'physics_suite'
+        >>> group = suite.find("group")
+        >>> group.attrib.get("name")
+        'main'
+        >>> group.find("scheme").text
+        'cloud_scheme'
+        >>> xml2 = '''
+        ... <suites>
+        ...   <suite name="physics_suite">
+        ...     <group name="main">
+        ...       <nested_suite name="microphysics_suite" group="micro"/>
+        ...     </group>
+        ...     <nested_suite name="pbl_suite"/>
+        ...   </suite>
+        ...   <suite name="microphysics_suite">
+        ...     <group name="micro">
+        ...       <scheme>cloud_scheme</scheme>
+        ...     </group>
+        ...   </suite>
+        ...   <suite name="pbl_suite">
+        ...     <nested_suite name="physics_suite"/>
+        ...   </suite>
+        ... </suites>
+        ... '''
+        >>> root2 = ET.fromstring(xml2)
+        >>> expand_nested_suites(root2, logger) #doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        ...
+        CCPPError: Infinite recursion while expanding nested suites: ['physics_suite', 'physics_suite']
+    """
+    # Keep track of any nested suites defined under the same root
+    # that need to be removed at the end of this function.
+    # This happens all in memory, it does not alter files on disk.
+    expanded_suites_to_remove = list()
+    # Iteratively expand nested suites until they are all gone
+    keep_expanding = True
+    while keep_expanding:
+        keep_expanding = False
+        ## To avoid infinite recursion, keep track of suite names
+        #suite_names = []
+        for suite in root.findall("suite"):
+            # To avoid infinite recursion, keep track of suite names
+            suite_names = [suite.attrib.get("name")]
+            # First, search all groups for nested_suite elements
+            groups = suite.findall("group")
+            for group in groups:
+                nested_suites = group.findall("nested_suite")
+                for nested in nested_suites:
+                    suite_name = replace_nested_suite(group, nested, root, default_path, logger)
+                    suite_names.append(suite_name)
+                    if not len(suite_names) == len(set(suite_names)):
+                        raise CCPPError(f"Infinite recursion while expanding nested suites: {suite_names}")
```

Comment:
Would it be more efficient to test first instead of two calls to `len` plus a `set` call?
```suggestion
                    if suite_name in suite_names:
                        raise CCPPError(f"Infinite recursion while expanding nested suites: {suite_names}")
                    # end if
                    suite_names.append(suite_name)
```

Reply:
My comment got removed I think; yes, I suggest collecting the suite names and spit them out together with the error message.

Reply:
Addressed in [28779432fae643cb600630559cbb8fbb388d68c9](https://github.com/NCAR/ccpp-framework/pull/691/commits/28779432fae643cb600630559cbb8fbb388d68c9). The output is now:
```
    parse_source.CCPPError: Exceeded number of iterations while expanding nested suites, check for inifite recursion or adjust limit max_iterations. Suites expanded so far: ['microphysics_suite', 'pbl_suite1', 'pbl_suite2', 'pbl_suite1', 'pbl_suite2', 'pbl_suite1', 'pbl_suite2', 'pbl_suite1', 'pbl_suite2', 'pbl_suite1', 'pbl_suite2']
```

Reply:
The original suggested code change is no longer valid and necessary because of the previous updates to count the number of iterations.